src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -6669,6 +6668,12 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                 throw SemanticError("str.join() takes one argument",
                     loc);
             }
+            ASR::expr_t *arg_sub = args[0].m_value;
+            ASR::ttype_t *arg_sub_type = ASRUtils::expr_type(arg_sub);
+            if(!ASR::is_a<ASR::List_t>(*arg_sub_type)){
+                throw SemanticError("str.join() takes type list only",
+                     loc);
+            }
```

Comment:
This change seems good. You can add an error reference test for this. Here is an example of how to add an error reference test https://github.com/lcompilers/lpython/commit/7c2bbac2e99c368383509de975a937ee813857ef. You just add a test file in `tests/errors`. In the test file, you pass a type apart from list to `str.join()`. You enable this test by adding it to `tests/tests.toml`. Then you just need to update the reference tests using `./run_tests.py -u`. This will add a new reference `.stderr` and `.json` files.

Reply:
Actually, I am still not checking the type of element inside the list which should be str, I checked grammar for asr but didn't get anything on how to iterate over list expression as it must be a list of str expressions. ast is getting generated perfectly but, I don't know how can I map ast nodes to asr nodes.

Reply:
Try updating refs using `./run_tests.py -u` and stage all and commit the updated refs.

Reply:
After implementing and committing the required changes, you can do the following to test everything on your machine:
```
./build0.sh
./build1.sh
./run_tests.py # if this fails, check for the failure and do ./run_tests.py -u to update refs.
cd integration_tests && ./run_tests.py # Here you can do `./run_tests.py -b c` to test c backend, use `-h` option to view all the available options
```
If everything works, then you are good to go.